torch/csrc/jit/codegen/cuda/fusion_segmenter.h

```diff
@@ -0,0 +1,380 @@
+#include <torch/csrc/jit/codegen/cuda/ir_base_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler_registry.h>
+
+#include <deque>
+#include <list>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class SegmentedGroup;
+class SegmentCandidateFinder;
+
+// A directed edge on DAG,
+// Wrapper for values, edges between segmented groups which are made up
+// of Exprs. Multiple edges can exist between segmented groups.
+struct SegmentedEdge {
+  SegmentedEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val)
+      : from(from), to(to), val(val) {}
+
+  SegmentedGroup* from;
+  SegmentedGroup* to;
+  Val* val;
+
+  void print();
+};
+
+std::ostream& operator<<(std::ostream& os, const SegmentedEdge* edge);
+
+//! Groups together expressions which create a segmented group
+//! Can be used to produce fusions
+class TORCH_CUDA_API SegmentedGroup {
+ public:
+  SegmentedGroup() = default;
+
+  SegmentedGroup(Expr* expr) {
+    exprs_.push_back(expr);
+  }
+
+  //! Checks if this group takes original fusion's input
+  bool isInputGroup() {
+    return !input_vals.empty();
+  };
+
+  //! Checks if this group is used any where in the segmented fusion
+  bool isConnected() const {
+    return !producer_edges.empty() || !consumer_edges.empty();
+  }
+
+  //! returns the id assigned by segment pass
+  int groupId() const {
+    return group_id_;
+  }
+
+  //! Returns inputs that this group shares with the original fusion
+  const auto& inputs() const {
+    return input_vals;
+  }
+
+  //! Returns outputs that this group shares with the original fusion
+  const auto& outputs() const {
+    return output_vals;
+  }
+
+  //! Returns the schedule heuristic associated with this group
+  ScheduleHeuristic heuristic() {
+    return heuristic_;
+  }
+
+  //! Returns the exprs that make up this group
+  const auto& exprs() const {
+    return exprs_;
+  }
+
+  //! Debug print function
+  void print();
+
+ public:
+  //! "Ancestor nodes", towards inputs of segmentedDAG
+  std::vector<SegmentedEdge*> producer_edges;
+
+  //! "Descendent nodes", towards outputs of segmentedDAG
+  std::vector<SegmentedEdge*> consumer_edges;
+
+  //! Composite Fusion inputs in this group
+  std::vector<Val*> input_vals;
+
+  //! Composite Fusion outputs in this group
+  std::vector<Val*> output_vals;
+
+ private:
+  friend class SegmentCandidateFinder;
+  friend class SegmentedFusion;
+  friend class FusionSegmentRuntime;
+
+  //! unique identifier of group in the segmented fusion
+  int group_id_ = -1;
+
+  //! The scheduler to use for compiling this group
+  ScheduleHeuristic heuristic_ = ScheduleHeuristic::PointWise;
+
+  //! Exprs that make up the group
+  std::vector<Expr*> exprs_;
+
+  //! Maximum path distance from an input segmented group required for
+  //! Theorem 4.2
+  int level = -1;
```

Comment:
the fields in this section still need _ suffix

Reply:
Updated, Thanks!